### PR TITLE
[JENKINS-52491] - PipelineGenerator extensibility and support of custom sections 

### PR DESCRIPTION
**Pull request by oleg-nenashev**
This is a result of the screenshare with @gautamabhishek46 and @martinda . We created a base extension engine and added support of custom pipeline sections defined as extensions. Also, some refactoring

- [x] - PipelineGenerator extension point
- [x] - Implementation for agent
- [x] - Custom sections support
- [x] - Detach all other generator code to extensions
- [ ] - Add annotations in methods
- [x] - Fix the exception propagation, use the new type
- [x] - Add extra unit tests

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/sprp/impl/AgentGenerator.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    public List<String> toPipeline(Agent agent) {
+        ArrayList<String> agentLines = new ArrayList<>();
+
+        if (agent == null) {
+            agentLines.add("agent any");
+        } else if (agent.getAnyOrNone() != null)
+            agentLines.add("agent " + agent.getAnyOrNone());
+        else if(agent.isNone()) {
```
Comment (kwhetstone) [inline src/main/java/io/jenkins/plugins/sprp/impl/AgentGenerator.java]: I'm still not sure why we have 2 checks here.  `getAnyOrNone()` literally returns `any` or `none`.  We should never reach this last check.

File: src/main/java/io/jenkins/plugins/sprp/impl/AgentGenerator.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+        return agentLines;
+    }
+
+    @Override
+    public boolean canConvert(Object object) {
+        return object instanceof Agent;
+    }
+
+    private List<String> getCommonOptionsOfAgent(Agent agent) {
```
Comment (kwhetstone) [inline src/main/java/io/jenkins/plugins/sprp/impl/AgentGenerator.java]: For getting this to work for now, this is good.  I think there's a way to get this from the extension point this matches.  I'd have to dig though; you could probably even ask on irc.

**Comment by oleg-nenashev** [comment]
There are still some bugs to be fixes:

```
[INFO] Total bugs: 3

[INFO] gitConfig must be non-null but is marked as nullable [io.jenkins.plugins.sprp.generators.GitPushStageGenerator] At GitPushStageGenerator.java:[lines 21-32] NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE

[INFO] reportsAndArtifactsInfo must be non-null but is marked as nullable [io.jenkins.plugins.sprp.generators.PublishReportsAndArtifactsStageGenerator] At PublishReportsAndArtifactsStageGenerator.java:[lines 24-54] NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE

[INFO] stage must be non-null but is marked as nullable [io.jenkins.plugins.sprp.generators.StageGenerator] At StageGenerator.java:[lines 29-54] NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE

[INFO] 

```

**Comment by kwhetstone** [comment]
Just a few general comments about the code that could be addressed in another PR.  There's a lot going on here.  I do really appreciate all the tests

**Comment by gautamabhishek46** [comment]
Not sure how to remove the findbugs error.

**Pull request merged**